### PR TITLE
Log missing images, but do not crash the game.

### PR DESCRIPTION
**Pull request by DanVanAtta**
Helps address:
#651 
#528

We still have an issue that unit images are invisible, but at least the game is still playable. Before this change we would get an uncaught exception and map rendering would stop, or the battle calc would not open, or the purchase units dialog would not open.

Some screenshots of what it looks like after this change:
__missing_marine
__missing_marine_edit
__missing_marine_production
__missing_maring_invisible_unit
missing_unit_move

---
This change is [Reviewable](https://reviewable.io/reviews/triplea-game/triplea/685)

### CODE REVIEW

**Comment by ron-murhammer** [comment]
@DanVanAtta Looks like this needs rebased.

**Comment by DanVanAtta** [comment]
merge conflicts resolved

**Comment by ron-murhammer** [comment]
@DanVanAtta Still showing conflicts to resolve for me. Might be some of the other PRs I just merged.

File: src/games/strategy/triplea/image/UnitImageFactory.java [hunk 1]
```diff
@@ -107,60 +111,77 @@ private void clearImageCache() {
...
-    // guaranteed to be loaded.
-    final int width = (int) (baseImage.getWidth(null) * m_scaleFactor);
-    final int height = (int) (baseImage.getHeight(null) * m_scaleFactor);
-    final Image scaledImage = baseImage.getScaledInstance(width, height, Image.SCALE_SMOOTH);
-    // Ensure the scaling is completed.
-    try {
-      Util.ensureImageLoaded(scaledImage);
-    } catch (final InterruptedException ex) {
-      ex.printStackTrace();
+    try (PerfTimer timer = Perf.startTimer("Load an image")) {
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/image/UnitImageFactory.java]: I'm guessing this is called a lot. Do you think its worth adding a timer here or is it too noisy?

File: src/games/strategy/triplea/oddsCalculator/ta/OddsCalculatorPanel.java [hunk 2]
```diff
@@ -1295,23 +1300,20 @@ private JPanel getUnitButtonPanel(final List<UnitCategory> categories, final JTe
...
-        });
-        panel.add(button);
+        final Optional<Image> img = m_context.getUnitImageFactory().getImage(category.getType(), category.getOwner(), m_data,
+            category.hasDamageOrBombingUnitDamage(), category.getDisabled());
+        if(img.isPresent()) {
+          final JButton button = new JButton(new ImageIcon(img.get()));
+          button.setToolTipText(toolTipText);
+          button.addActionListener(e -> textField
+              .setText((textField.getText().length() > 0 ? (textField.getText() + OddsCalculator.OOL_SEPARATOR) : "")
+                  + unitName));
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/oddsCalculator/ta/OddsCalculatorPanel.java]: Do you think it would be better to break this line up some to improve readability?

File: src/games/strategy/triplea/ui/MapPanel.java [hunk 3]
```diff
@@ -563,13 +564,15 @@ public void paint(final Graphics g) {
...
-          final BufferedImage highlight =
-              (BufferedImage) uiContext.getUnitImageFactory().getHighlightImage(category.getType(),
-                  category.getOwner(), m_data, category.hasDamageOrBombingUnitDamage(), category.getDisabled());
-          final AffineTransform t = new AffineTransform();
-          t.translate(normalizeX(r.getX() - getXOffset()) * m_scale, normalizeY(r.getY() - getYOffset()) * m_scale);
-          t.scale(m_scale, m_scale);
-          g2d.drawImage(highlight, t, this);
+
+          Optional<Image> image = uiContext.getUnitImageFactory().getHighlightImage(category.getType(),
+              category.getOwner(), m_data, category.hasDamageOrBombingUnitDamage(), category.getDisabled());;
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/MapPanel.java]: Double semi-colon

File: src/games/strategy/triplea/ui/ProductionPanel.java [hunk 4]
```diff
@@ -284,7 +275,7 @@ protected JPanel getPanelComponent() {
       } else {
         text = "<html> x " + ResourceCollection.toStringForHTML(m_cost, m_data) + "</html>";
       }
-      final JLabel label = new JLabel(text, icon, SwingConstants.LEFT);
+      final JLabel label = icon.isPresent() ? new JLabel(text, icon.get(), SwingConstants.LEFT) : new JLabel( text, SwingConstants.LEFT);
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/ProductionPanel.java]: This line looks too long. Did you format?

File: src/games/strategy/triplea/ui/UIContext.java [hunk 5]
```diff
@@ -132,6 +136,17 @@ public UnitImageFactory getUnitImageFactory() {
     return m_unitImageFactory;
   }
 
+  @Override public JLabel createUnitImageJLabel(UnitType type, PlayerID player, GameData data,
+      UnitDamage damaged, UnitEnable disabled) {
+    Optional<ImageIcon>
+        image = this.getUnitImageFactory().getIcon(type,player,data,damaged == UnitDamage.DAMAGED, disabled == UnitEnable.DISABLED);
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/UIContext.java]: Could use some spaces between parameters. Also is the 'this' needed?

File: src/games/strategy/triplea/ui/UIContext.java [hunk 6]
```diff
@@ -132,6 +136,17 @@ public UnitImageFactory getUnitImageFactory() {
     return m_unitImageFactory;
   }
 
+  @Override public JLabel createUnitImageJLabel(UnitType type, PlayerID player, GameData data,
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/UIContext.java]: `@Override` should probably be on a separate line.

File: src/games/strategy/triplea/ui/screen/UnitsDrawer.java [hunk 7]
```diff
@@ -80,44 +81,53 @@ public void draw(final Rectangle bounds, final GameData data, final Graphics2D g
...
-          //currently the flag is drawn in the lower right corner of the image's bounds -> offsets on some unit images
-          //This Method draws the Flag in the lower right corner of the unit image. Since the position is the upper left corner we have to move the picture up by the height and left by the width.
-          graphics.drawImage(flag, (placementPoint.x - bounds.x) + xoffset, (placementPoint.y - bounds.y) + yoffset, null);
-        } 
-        break;
-      case NONE:
-        //This Method draws the unit Image
-        graphics.drawImage(img, placementPoint.x - bounds.x, placementPoint.y - bounds.y, null);
-        break;
+    final Optional<Image> img = uiContext.getUnitImageFactory().getImage(type, owner, data, damaged > 0 || bombingUnitDamage > 0, disabled);
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/screen/UnitsDrawer.java]: This class probably needs formatted as there are a few long lines here.

File: src/games/strategy/triplea/image/UnitImageFactory.java (hunk 1 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/image/UnitImageFactory.java]: The timer is good for spot checking stuff, but client side timing I'm rapidly coming to the opinion is without value. Will remove.

File: src/games/strategy/triplea/image/UnitImageFactory.java [hunk 8]
```diff
@@ -107,60 +111,77 @@ private void clearImageCache() {
...
 
-  public static URL getBaseImageURL(final String baseImageName, final PlayerID id,
+  private static Optional<URL> getBaseImageURL(final String baseImageName, final PlayerID id,
       final ResourceLoader resourceLoader) {
     // URL uses '/' not '\'
     final String fileName = FILE_NAME_BASE + id.getName() + "/" + baseImageName + ".png";
     final URL url = resourceLoader.getResource(fileName);
     if (url == null) {
-      throw new IllegalStateException("Cant load: " + baseImageName + "  looking in: " + fileName);
+      ClientLogger.logError("MISSING IMAGE (this unit or image will be invisible): " + baseImageName + "  looking in: " + fileName);
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/image/UnitImageFactory.java]: Should this error be logged at a higher level? Probably want to make sure the user is aware of the problem since the map is very difficult to play and its confusing when images are missing.

Comment (DanVanAtta) [inline src/games/strategy/triplea/image/UnitImageFactory.java]: The logError call will throw a pop-up informing the user.  This is the right place to log it, when we introduce the Optional.empty() in the first place. There are multiple independent higher levels that call into this code, they all would have to do the same thing if we logged the problem there.

File: src/games/strategy/triplea/oddsCalculator/ta/OddsCalculatorPanel.java (hunk 2 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/oddsCalculator/ta/OddsCalculatorPanel.java]: Yes, but.. too much scope creep.  Getting rid of the anonymous inner class was a quick and easy win. Breaking up the line opens a can of worms.

**Comment by DanVanAtta** [comment]
Updated

Note: should be squash merged, remove all comments except the first
